File: examples/java-flaky-ci-fixture/CheckoutTest.java
// Planted order-dependent test failure that fires only under CI-like
// conditions, for the flaky-CI wedge (Track 3), Java port of
// examples/flaky-ci-fixture.
//
// The first test runs ONLY on the CI legacy matrix (CI_LEGACY_MATRIX=1) and
// leaks state into the shared config service: it switches the service to its
// legacy response format, which returns the tax rate as a string. The second
// test then computes a wrong total and fails. A plain local run never takes
// the legacy branch, so the suite passes and the failure looks
// unreproducible ("flaky"). The capsule spooled by the CI run carries the
// recorded legacy response, so `reproit check <capsule> --exec "java -cp
// classes CheckoutTest"` re-executes the exact failing run anywhere.
//
// Uses the SDK's dependency-free Ci micro-runner (the JUnit-shaped surface
// is ReproitCi) so replay compiles with plain javac and needs no jars and no
// network. Compile:
//   javac -d classes -sourcepath sdk/reproit-backend-java/src/main/java \
//     examples/java-flaky-ci-fixture/CheckoutTest.java

import com.sun.net.httpserver.HttpServer;
import dev.reproit.backend.Ci;
import dev.reproit.backend.Instrument;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.nio.charset.StandardCharsets;

public final class CheckoutTest {
    private static final int PORT = 19992;
    private static final String CONFIG_URL = "http://127.0.0.1:" + PORT;

    // The shared config service both tests talk to. Stateful on purpose: the
    // legacy-format test leaks its toggle into it. Never started under
    // replay, where the SDK serves the recorded exchanges in process and any
    // real socket attempt would surface as a divergence, not a connection.
    private static boolean legacy = false;

    private CheckoutTest() {}

    private static HttpServer configService() throws Exception {
        HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", PORT), 0);
        server.createContext("/", exchange -> {
            if ("POST".equals(exchange.getRequestMethod())
                    && "/format/legacy".equals(exchange.getRequestURI().getPath())) {
                legacy = true;
                exchange.sendResponseHeaders(204, -1);
                exchange.close();
                return;
            }
            byte[] body = (legacy ? "{\"rate\":\"0.25\"}" : "{\"rate\":0.25}")
                .getBytes(StandardCharsets.UTF_8);
            exchange.getResponseHeaders().set("Content-Type", "application/json");
            exchange.sendResponseHeaders(200, body.length);
            exchange.getResponseBody().write(body);
            exchange.close();
        });
        server.start();
        return server;
    }

    public static void main(String[] args) throws Exception {
        Instrument.init();
        HttpServer server =
            System.getenv("REPROIT_REPLAY") == null ? configService() : null;
        HttpClient client = HttpClient.newHttpClient();
        Ci.Suite suite = Ci.suite("checkout");

        suite.test("legacy config format toggles", () -> {
            // CI-only: this is the state leak that makes the next test order
            // dependent. A local run never takes this branch.
            if (!"1".equals(System.getenv("CI_LEGACY_MATRIX"))) return;
            Instrument.Http.ExchangeResponse response = Instrument.Http.send(
                client,
                HttpRequest.newBuilder(URI.create(CONFIG_URL + "/format/legacy"))
                    .POST(HttpRequest.BodyPublishers.noBody()).build());
            if (response.status() != 204) {
                throw new AssertionError("legacy toggle answered " + response.status());
            }
        });

        suite.test("order total applies the configured tax rate", () -> {
            long total = Order.orderTotal(100, CONFIG_URL, client);
            if (total != 125) {
                throw new AssertionError("order total expected 125, got " + total);
            }
        });

        if (server != null) server.stop(0);
        System.exit(suite.exitCode());
    }
}

File: fixtures/php-flaky-ci-fixture/tests/checkout_test.php
<?php

// Planted order-dependent test failure that fires only under CI-like
// conditions, for the flaky-CI wedge (Track 3), PHP edition.
//
// The first test runs ONLY on the CI legacy matrix (CI_LEGACY_MATRIX=1) and
// leaks state into the shared config service: it switches the service to
// its legacy response format, which wraps the tax rate. The second test
// then computes a wrong total and fails. A plain local run never takes the
// legacy branch, so the suite passes and the failure looks unreproducible
// ("flaky"). The capsule spooled by the CI run carries the recorded legacy
// response, so `reproit check <capsule> --exec "php tests/checkout_test.php"`
// re-executes the exact failing run anywhere.
//
// Run it directly (`php tests/checkout_test.php`): the SDK's plain-script
// CI wrapper (Ci::suite) owns the exit code and the stderr markers
// `reproit check` parses.

declare(strict_types=1);

require __DIR__ . '/../../../sdk/reproit-backend-php/reproit.php';
require __DIR__ . '/../../../sdk/reproit-backend-php/ci.php';
require __DIR__ . '/../order.php';

use ReproitBackend\Ci;
use ReproitBackend\Instrument;

use function PhpFlakyCiFixture\order_total;

const PORT = 19991;
$configUrl = 'http://127.0.0.1:' . PORT;

// The shared config service both tests talk to. Never started under replay,
// where the SDK serves the recorded exchanges in process and any real
// socket attempt would surface as a divergence, not a connection.
$replay = getenv('REPROIT_REPLAY');
if (!\is_string($replay) || $replay === '') {
    $state = sys_get_temp_dir() . '/reproit-php-flaky-' . getmypid();
    @mkdir($state, 0o777, true);
    $spec = [1 => ['file', '/dev/null', 'w'], 2 => ['file', '/dev/null', 'w']];
    $service = proc_open(
        [PHP_BINARY, '-S', '127.0.0.1:' . PORT, __DIR__ . '/../config_service.php'],
        $spec,
        $pipes,
        null,
        ['REPROIT_FIXTURE_STATE' => $state] + getenv()
    );
    if ($service === false) {
        fwrite(STDERR, "config service failed to start\n");
        exit(2);
    }
    $ready = false;
    for ($i = 0; $i < 100 && !$ready; $i++) {
        // Raw readiness probe, deliberately outside the SDK boundary: no
        // ambient trace exists yet, so nothing is recorded.
        $ready = @file_get_contents($configUrl . '/tax-rate') !== false;
        if (!$ready) {
            usleep(50000);
        }
    }
    register_shutdown_function(static function () use ($service, $state): void {
        @proc_terminate($service);
        @unlink($state . '/legacy.flag');
        @rmdir($state);
    });
    if (!$ready) {
        fwrite(STDERR, "config service never answered\n");
        exit(2);
    }
}

$test = Ci::suite('checkout');

$test('legacy config format toggles', function () use ($configUrl): void {
    // CI-only: this is the state leak that makes the next test order
    // dependent. A local run never takes this branch.
    if (getenv('CI_LEGACY_MATRIX') !== '1') {
        return;
    }
    $response = Instrument::http('POST', $configUrl . '/format/legacy');
    if ($response->status !== 204) {
        throw new \RuntimeException('legacy toggle answered ' . $response->status);
    }
});

$test('order total applies the configured tax rate', function () use ($configUrl): void {
    $total = order_total(100, $configUrl);
    if ($total !== 125.0) {
        throw new \RuntimeException('order total ' . $total . ' does not equal 125');
    }
});
